core/src/main/scala/org/http4s/package.scala

```diff
@@ -82,4 +84,54 @@ package object http4s {
   type Http4sSyntax = syntax.AllSyntax
   @deprecated("Moved to org.http4s.syntax.all", "0.16")
   val Http4sSyntax = syntax.all
+
+  // Helper functions for writing Order instances //
+
+  /** This is the same as `Order.by(f).compare(a, b)`, but with the parameters
+    * re-arraigned to make it easier to partially apply the function to two
+    * instances of a type before supplying the `A => B`.
+    *
+    * The intended use case is that `f: A => B` will extract out a single
+    * field from two instances of some Product type and then compare the value
+    * of the field. This can then be done in turn for each field of a Product,
+    * significantly reducing the amount of code needed to write an `Order`
+    * instance for a Product with many fields.
+    *
+    * See the `Order` instance for `Uri` for an example of this usage.
+    */
+  private[http4s] def compareField[A, B: Order](
+      a: A,
+      b: A,
+      f: A => B
+  ): Int =
+    Order.by[A, B](f).compare(a, b)
+
+  /** Given at least one `Int` intended to represent the result of a comparison
+    * of two fields of some Product type, reduce the result to the first
+    * non-zero value, or return 0 if all comparisons are 0.
+    *
+    * The intended use case for this function is to reduce the amount of code
+    * needed to write an `Order` instance for Product types. One can use
+    * [[#compareField]] to generate a comparison for each field in a product
+    * type, then apply this function to get a ordering for the entire Product
+    * type.
+    *
+    * See the `Order` instance for `Uri` for an example of this usage.
+    *
+    * @note It is important to remember that `reduceMap` and `reduceMapK`
+    *       produce different results, since `Semigroup[Option[A]].combine`
+    *       and `SemigroupK[Option].combineK` have different semantics. We
+    *       ''always'' want the `SemigroupK[Option]` instance here.
+    */
+  private[http4s] def reduceComparisons(
```

Comment:
This runs all the comparisons, right?  Could we do a lazy one that only continues for as long as we get zeroes?